File: sdk/assemblyscript/src/assembly/collections.ts
/*
 * Copyright 2024 Hypermode Inc.
 * Licensed under the terms of the Apache License, Version 2.0
 * See the LICENSE file that accompanied this code for further details.
 *
 * SPDX-FileCopyrightText: 2024 Hypermode Inc. <hello@hypermode.com>
 * SPDX-License-Identifier: Apache-2.0
 */

import * as utils from "./utils";

export type CollectionStatus = string;
// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace CollectionStatus {
  export const Success = "success";
  export const Error = "error";
}
abstract class CollectionResult {
  collection: string;
  status: CollectionStatus;
  error: string;
  get isSuccessful(): bool {
    return this.status == CollectionStatus.Success;
  }

  constructor(collection: string, status: CollectionStatus, error: string) {
    this.collection = collection;
    this.status = status;
    this.error = error;
  }
}
export class CollectionMutationResult extends CollectionResult {
  operation: string;
  keys: string[] = [];

  constructor(
    collection: string,
    status: CollectionStatus,
    error: string,
    operation: string,
  ) {
    super(collection, status, error);
    this.operation = operation;
  }
}
export class SearchMethodMutationResult extends CollectionResult {
  operation: string;
  searchMethod: string;

  constructor(
    collection: string,
    status: CollectionStatus,
    error: string,
    operation: string,
    searchMethod: string,
  ) {
    super(collection, status, error);
    this.operation = operation;
    this.searchMethod = searchMethod;
  }
}
export class CollectionSearchResult extends CollectionResult {
  searchMethod: string;
  objects: CollectionSearchResultObject[];

  constructor(
    collection: string,
    status: CollectionStatus,
    error: string,
    searchMethod: string,
    objects: CollectionSearchResultObject[],
  ) {
    super(collection, status, error);
    this.searchMethod = searchMethod;
    this.objects = objects;
  }
}

export class CollectionSearchResultObject {
  namespace: string;
  key: string;
  text: string;
  labels: string[];
  distance: f64;
  score: f64;

  constructor(
    namespace: string,
    key: string,
    text: string,
    labels: string[],
    distance: f64,
    score: f64,
  ) {
    this.namespace = namespace;
    this.key = key;
    this.text = text;
    this.labels = labels;
    this.distance = distance;
    this.score = score;
  }
}

export class CollectionClassificationResult extends CollectionResult {
  searchMethod: string;
  labelsResult: CollectionClassificationLabelObject[];
  cluster: CollectionClassificationResultObject[];

  constructor(
    collection: string,
    status: CollectionStatus,
    error: string,
    searchMethod: string,
    labelsResult: CollectionClassificationLabelObject[],
    cluster: CollectionClassificationResultObject[],
  ) {
    super(collection, status, error);
    this.searchMethod = searchMethod;
    this.labelsResult = labelsResult;
    this.cluster = cluster;
  }
}

export class CollectionClassificationLabelObject {
  label: string;
  confidence: f64;

  constructor(label: string, confidence: f64) {
    this.label = label;
    this.confidence = confidence;
  }
}

export class CollectionClassificationResultObject {
  key: string;
  labels: string[];
  distance: f64;
  score: f64;

  constructor(key: string, labels: string[], distance: f64, score: f64) {
    this.key = key;
    this.labels = labels;
    this.distance = distance;
    this.score = score;
  }
}

// @ts-expect-error: decorator
@external("modus_collections", "upsert")
declare function hostUpsertToCollection(
  collection: string,
  namespace: string,
  keys: string[],
  texts: string[],
  labels: string[][],
): CollectionMutationResult;

// @ts-expect-error: decorator
@external("modus_collections", "delete")
declare function hostDeleteFromCollection(
  collection: string,
  namespace: string,
  key: string,
): CollectionMutationResult;

// @ts-expect-error: decorator
@external("modus_collections", "search")
declare function hostSearchCollection(
  collection: string,
  namespaces: string[],
  searchMethod: string,
  text: string,
  limit: i32,
  returnText: bool,
): CollectionSearchResult;

// @ts-expect-error: decorator
@external("modus_collections", "classifyText")
declare function hostNnClassifyCollection(
  collection: string,
  namespace: string,
  searchMethod: string,
  text: string,
): CollectionClassificationResult;

// @ts-expect-error: decorator
@external("modus_collections", "recomputeIndex")
declare function hostRecomputeSearchMethod(
  collection: string,
  namespace: string,
  searchMethod: string,
): SearchMethodMutationResult;

// @ts-expect-error: decorator
@external("modus_collections", "computeDistance")
declare function hostComputeDistance(
  collection: string,
  namespace: string,
  searchMethod: string,
  key1: string,
  key2: string,
): CollectionSearchResultObject;

// @ts-expect-error: decorator
@external("modus_collections", "getText")
declare function hostGetTextFromCollection(
  collection: string,
  namespace: string,
  key: string,
): string;

// @ts-expect-error: decorator
@external("modus_collections", "dumpTexts")
declare function hostGetTextsFromCollection(
  collection: string,
  namespace: string,
): Map<string, string>;

// @ts-expect-error: decorator
@external("modus_collections", "getNamespaces")
declare function hostGetNamespacesFromCollection(collection: string): string[];

// @ts-expect-error: decorator
@external("modus_collections", "getVector")
declare function hostGetVector(
  collection: string,
  namespace: string,
  searchMethod: string,
  key: string,
): f32[];

// @ts-expect-error: decorator
@external("modus_collections", "getLabels")
declare function hostGetLabels(
  collection: string,
  namespace: string,
  key: string,
): string[];

// @ts-expect-error: decorator
@external("modus_collections", "searchByVector")
declare function hostSearchCollectionByVector(
  collection: string,
  namespaces: string[],
  searchMethod: string,
  vector: f32[],
  limit: i32,
  returnText: bool,
): CollectionSearchResult;

// add batch upsert
export function upsertBatch(
  collection: string,
  keys: string[] | null,
  texts: string[],
  labelsArr: string[][] = [],
  namespace: string = "",
): CollectionMutationResult {
  if (collection.length == 0) {
    console.error("Collection is empty.");
    return new CollectionMutationResult(
      collection,
      CollectionStatus.Error,
      "Collection is empty.",
      "upsert",
    );
  }
  if (texts.length == 0) {
    console.error("Texts is empty.");
    return new CollectionMutationResult(
      collection,
      CollectionStatus.Error,
      "Texts is empty.",
      "upsert",
    );
  }
  let keysArr: string[] = [];
  if (keys != null) {
    keysArr = keys;
  }

  const result = hostUpsertToCollection(
    collection,
    namespace,
    keysArr,
    texts,
    labelsArr,
  );
  if (utils.resultIsInvalid(result)) {
    console.error("Error upserting to Text index.");
    return new CollectionMutationResult(
      collection,
      CollectionStatus.Error,
      "Error upserting to Text index.",
      "upsert",
    );
  }
  return result;
}

// add data to in-mem storage, get all embedders for a collection, run text through it
// and insert the Text into the Text indexes for each search method
export function upsert(
  collection: string,
  key: string | null,
  text: string,
  labels: string[] = [],
  namespace: string = "",
): CollectionMutationResult {
  if (collection.length == 0) {
    console.error("Collection is empty.");
    return new CollectionMutationResult(
      collection,
      CollectionStatus.Error,
      "Collection is empty.",
      "upsert",
    );
  }
  if (text.length == 0) {
    console.error("Text is empty.");
    return new CollectionMutationResult(
      collection,
      CollectionStatus.Error,
      "Text is empty.",
      "upsert",
    );
  }
  const keys: string[] = [];
  if (key != null) {
    keys.push(key);
  }

  const texts: string[] = [text];

  const labelsArr: string[][] = [labels];

  const result = hostUpsertToCollection(
    collection,
    namespace,
    keys,
    texts,
    labelsArr,
  );
  if (utils.resultIsInvalid(result)) {
    console.error("Error upserting to Text index.");
    return new CollectionMutationResult(
      collection,
      CollectionStatus.Error,
      "Error upserting to Text index.",
      "upsert",
    );
  }
  return result;
}

// remove data from in-mem storage and indexes
export function remove(
  collection: string,
  key: string,
  namespace: string = "",
): CollectionMutationResult {
  if (collection.length == 0) {
    console.error("Collection is empty.");
    return new CollectionMutationResult(
      collection,
      CollectionStatus.Error,
      "Collection is empty.",
      "delete",
    );
  }
  if (key.length == 0) {
    console.error("Key is empty.");
    return new CollectionMutationResult(
      collection,
      CollectionStatus.Error,
      "Key is empty.",
      "delete",
    );
  }
  const result = hostDeleteFromCollection(collection, namespace, key);
  if (utils.resultIsInvalid(result)) {
    console.error("Error deleting from Text index.");
    return new CollectionMutationResult(
      collection,
      CollectionStatus.Error,
      "Error deleting from Text index.",
      "delete",
    );
  }
  return result;
}

// fetch embedders for collection & search method, run text through it and
// search Text index for similar Texts, return the result keys
// open question: how do i return a more expansive result from string array
export function search(
  collection: string,
  searchMethod: string,
  text: string,
  limit: i32,
  returnText: bool = false,
  namespaces: string[] = [],
): CollectionSearchResult {
  if (text.length == 0) {
    return new CollectionSearchResult(
      collection,
      CollectionStatus.Error,
      "Text is empty.",
      searchMethod,
      [],
    );
  }
  const result = hostSearchCollection(
    collection,
    namespaces,
    searchMethod,
    text,
    limit,
    returnText,
  );
  if (utils.resultIsInvalid(result)) {
    console.error("Error searching Text index.");
    return new CollectionSearchResult(
      collection,
      CollectionStatus.Error,
      "Error searching Text index.",
      searchMethod,
      [],
    );
  }
  return result;
}

export function searchByVector(
  collection: string,
  searchMethod: string,
  vector: f32[],
  limit: i32,
  returnText: bool = false,
  namespaces: string[] = [],
): CollectionSearchResult {
  if (vector.length == 0) {
    return new CollectionSearchResult(
      collection,
      CollectionStatus.Error,
      "Vector is empty.",
      searchMethod,
      [],
    );
  }
  const result = hostSearchCollectionByVector(
    collection,
    namespaces,
    searchMethod,
    vector,
    limit,
    returnText,
  );
  if (utils.resultIsInvalid(result)) {
    console.error("Error searching Text index by vector.");
    return new CollectionSearchResult(
      collection,
      CollectionStatus.Error,
      "Error searching Text index by vector.",
      searchMethod,
      [],
    );
  }
  return result;
}

// fetch embedders for collection & search method, run text through it and
// classify Text index for similar Texts, return the result keys
export function nnClassify(
  collection: string,
  searchMethod: string,
  text: string,
  namespace: string = "",
): CollectionClassificationResult {
  if (text.length == 0) {
    return new CollectionClassificationResult(
      collection,
      CollectionStatus.Error,
      "Text is empty.",
      searchMethod,
      [],
      [],
    );
  }
  const result = hostNnClassifyCollection(
    collection,
    namespace,
    searchMethod,
    text,
  );
  if (utils.resultIsInvalid(result)) {
    console.error("Error classifying Text index.");
    return new CollectionClassificationResult(
      collection,
      CollectionStatus.Error,
      "Error classifying Text index.",
      searchMethod,
      [],
      [],
    );
  }
  return result;
}

export function recomputeSearchMethod(
  collection: string,
  searchMethod: string,
  namespace: string = "",
): SearchMethodMutationResult {
  if (collection.length == 0) {
    console.error("Collection is empty.");
    return new SearchMethodMutationResult(
      collection,
      CollectionStatus.Error,
      "Collection is empty.",
      "recompute",
      searchMethod,
    );
  }
  if (searchMethod.length == 0) {
    console.error("Search method is empty.");
    return new SearchMethodMutationResult(
      collection,
      CollectionStatus.Error,
      "Search method is empty.",
      "recompute",
      searchMethod,
    );
  }
  const result = hostRecomputeSearchMethod(collection, namespace, searchMethod);
  if (utils.resultIsInvalid(result)) {
    console.error("Error recomputing Text index.");
    return new SearchMethodMutationResult(
      collection,
      CollectionStatus.Error,
      "Error recomputing Text index.",
      "recompute",
      searchMethod,
    );
  }
  return result;
}

/**
 * @deprecated Use `collections.computeDistance` instead.
 */
export function computeSimilarity(
  collection: string,
  searchMethod: string,
  key1: string,
  key2: string,
  namespace: string = "",
): CollectionSearchResultObject {
  return computeDistance(collection, searchMethod, key1, key2, namespace);
}

export function computeDistance(
  collection: string,
  searchMethod: string,
  key1: string,
  key2: string,
  namespace: string = "",
): CollectionSearchResultObject {
  if (collection.length == 0) {
    console.error("Collection is empty.");
    return new CollectionSearchResultObject("", "", "", [], 0.0, 0.0);
  }
  if (searchMethod.length == 0) {
    console.error("Search method is empty.");
    return new CollectionSearchResultObject("", "", "", [], 0.0, 0.0);
  }
  if (key1.length == 0) {
    console.error("Key1 is empty.");
    return new CollectionSearchResultObject("", "", "", [], 0.0, 0.0);
  }
  if (key2.length == 0) {
    console.error("Key2 is empty.");
    return new CollectionSearchResultObject("", "", "", [], 0.0, 0.0);
  }
  return hostComputeDistance(collection, namespace, searchMethod, key1, key2);
}

export function getText(
  collection: string,
  key: string,
  namespace: string = "",
): string {
  if (collection.length == 0) {
    console.error("Collection is empty.");
    return "";
  }
  if (key.length == 0) {
    console.error("Key is empty.");
    return "";
  }
  return hostGetTextFromCollection(collection, namespace, key);
}

export function getTexts(
  collection: string,
  namespace: string = "",
): Map<string, string> {
  if (collection.length == 0) {
    console.error("Collection is empty.");
    return new Map<string, string>();
  }
  return hostGetTextsFromCollection(collection, namespace);
}

export function getNamespaces(collection: string): string[] {
  if (collection.length == 0) {
    console.error("Collection is empty.");
    return [];
  }
  return hostGetNamespacesFromCollection(collection);
}

export function getVector(
  collection: string,
  searchMethod: string,
  key: string,
  namespace: string = "",
): f32[] {
  if (collection.length == 0) {
    console.error("Collection is empty.");
    return [];
  }
  if (searchMethod.length == 0) {
    console.error("Search method is empty.");
    return [];
  }
  if (key.length == 0) {
    console.error("Key is empty.");
    return [];
  }
  return hostGetVector(collection, namespace, searchMethod, key);
}

export function getLabels(
  collection: string,
  key: string,
  namespace: string = "",
): string[] {
  if (collection.length == 0) {
    console.error("Collection is empty.");
    return [];
  }
  if (key.length == 0) {
    console.error("Key is empty.");
    return [];
  }
  return hostGetLabels(collection, namespace, key);
}

File: sdk/go/pkg/collections/collections.go
/*
 * Copyright 2025 Hypermode Inc.
 * Licensed under the terms of the Apache License, Version 2.0
 * See the LICENSE file that accompanied this code for further details.
 *
 * SPDX-FileCopyrightText: 2025 Hypermode Inc. <hello@hypermode.com>
 * SPDX-License-Identifier: Apache-2.0
 */

package collections

import (
	"fmt"
)

type CollectionStatus = string

const (
	Success CollectionStatus = "success"
	Error   CollectionStatus = "error"
)

type CollectionMutationResult struct {
	Collection string
	Status     string
	Error      string
	Operation  string
	Keys       []string
}

type SearchMethodMutationResult struct {
	Collection   string
	Status       string
	Error        string
	Operation    string
	SearchMethod string
}

type CollectionSearchResult struct {
	Collection   string
	Status       string
	Error        string
	SearchMethod string
	Objects      []*CollectionSearchResultObject
}

type CollectionSearchResultObject struct {
	Namespace string
	Key       string
	Text      string
	Labels    []string
	Distance  float64
	Score     float64
}

type CollectionClassificationResult struct {
	Collection   string
	Status       string
	Error        string
	SearchMethod string
	LabelsResult []*CollectionClassificationLabelObject
	Cluster      []*CollectionClassificationResultObject
}

type CollectionClassificationLabelObject struct {
	Label      string
	Confidence float64
}

type CollectionClassificationResultObject struct {
	Key      string
	Labels   []string
	Distance float64
	Score    float64
}

type NamespaceOption func(*NamespaceOptions)

type NamespaceOptions struct {
	namespace string
}

func WithNamespace(namespace string) NamespaceOption {
	return func(o *NamespaceOptions) {
		o.namespace = namespace
	}
}

func UpsertBatch(collection string, keys []string, texts []string, labelsArr [][]string, opts ...NamespaceOption) (*CollectionMutationResult, error) {
	if collection == "" {
		return nil, fmt.Errorf("Collection name is required")
	}

	if len(texts) == 0 {
		return nil, fmt.Errorf("Texts is empty")
	}

	nsOpts := &NamespaceOptions{
		namespace: "",
	}

	for _, opt := range opts {
		opt(nsOpts)
	}

	if keys == nil {
		keys = []string{}
	}

	if labelsArr == nil {
		labelsArr = [][]string{}
	}

	result := hostUpsert(&collection, &nsOpts.namespace, &keys, &texts, &labelsArr)

	if result == nil {
		return nil, fmt.Errorf("Failed to upsert")
	}

	return result, nil
}

func Upsert(collection string, key *string, text string, labels []string, opts ...NamespaceOption) (*CollectionMutationResult, error) {
	if collection == "" {
		return nil, fmt.Errorf("Collection name is required")
	}

	keyArr := []string{}

	if key != nil {
		keyArr = []string{*key}
	}

	labelsArr := [][]string{}

	if labels != nil {
		labelsArr = [][]string{labels}
	}

	if text == "" {
		return nil, fmt.Errorf("Text is required")
	}

	nsOpts := &NamespaceOptions{
		namespace: "",
	}

	for _, opt := range opts {
		opt(nsOpts)
	}

	result := hostUpsert(&collection, &nsOpts.namespace, &keyArr, &[]string{text}, &labelsArr)

	if result == nil {
		return nil, fmt.Errorf("Failed to upsert")
	}

	return result, nil
}

func Remove(collection, key string, opts ...NamespaceOption) (*CollectionMutationResult, error) {
	if collection == "" {
		return nil, fmt.Errorf("Collection name is required")
	}

	if key == "" {
		return nil, fmt.Errorf("Key is required")
	}

	nsOpts := &NamespaceOptions{
		namespace: "",
	}

	for _, opt := range opts {
		opt(nsOpts)
	}

	result := hostDelete(&collection, &nsOpts.namespace, &key)

	if result == nil {
		return nil, fmt.Errorf("Failed to delete")
	}

	return result, nil
}

type SearchOption func(*SearchOptions)

type SearchOptions struct {
	namespaces []string
	limit      int
	returnText bool
}

func WithNamespaces(namespaces []string) SearchOption {
	return func(o *SearchOptions) {
		o.namespaces = namespaces
	}
}

func WithLimit(limit int) SearchOption {
	return func(o *SearchOptions) {
		o.limit = limit
	}
}

func WithReturnText(returnText bool) SearchOption {
	return func(o *SearchOptions) {
		o.returnText = returnText
	}
}

func Search(collection, searchMethod, text string, opts ...SearchOption) (*CollectionSearchResult, error) {
	if collection == "" {
		return nil, fmt.Errorf("Collection name is required")
	}

	if searchMethod == "" {
		return nil, fmt.Errorf("Search method is required")
	}

	if text == "" {
		return nil, fmt.Errorf("Text is required")
	}

	sOpts := &SearchOptions{
		namespaces: []string{},
		limit:      10,
		returnText: false,
	}

	for _, opt := range opts {
		opt(sOpts)
	}

	result := hostSearch(&collection, &sOpts.namespaces, &searchMethod, &text, int32(sOpts.limit), sOpts.returnText)

	if result == nil {
		return nil, fmt.Errorf("Failed to search")
	}

	return result, nil
}

func SearchByVector(collection, searchMethod string, vector []float32, opts ...SearchOption) (*CollectionSearchResult, error) {
	if collection == "" {
		return nil, fmt.Errorf("Collection name is required")
	}

	if searchMethod == "" {
		return nil, fmt.Errorf("Search method is required")
	}

	if len(vector) == 0 {
		return &CollectionSearchResult{
			Collection:   collection,
			Status:       Error,
			Error:        "Vector is required",
			SearchMethod: "",
			Objects:      nil,
		}, fmt.Errorf("Vector is required")
	}

	sOpts := &SearchOptions{
		namespaces: []string{},
		limit:      10,
		returnText: false,
	}

	for _, opt := range opts {
		opt(sOpts)
	}

	result := hostSearchByVector(&collection, &sOpts.namespaces, &searchMethod, &vector, int32(sOpts.limit), sOpts.returnText)

	if result == nil {
		return nil, fmt.Errorf("Failed to search")
	}

	return result, nil
}

func NnClassify(collection, searchMethod, text string, opts ...NamespaceOption) (*CollectionClassificationResult, error) {
	if collection == "" {
		return nil, fmt.Errorf("Collection name is required")
	}

	if searchMethod == "" {
		return nil, fmt.Errorf("Search method is required")
	}

	if text == "" {
		return nil, fmt.Errorf("Text is required")
	}

	nsOpts := &NamespaceOptions{
		namespace: "",
	}

	for _, opt := range opts {
		opt(nsOpts)
	}

	result := hostClassifyText(&collection, &nsOpts.namespace, &searchMethod, &text)

	if result == nil {
		return nil, fmt.Errorf("Failed to classify")
	}

	return result, nil
}

func RecomputeSearchMethod(collection, searchMethod string, opts ...NamespaceOption) (*SearchMethodMutationResult, error) {
	if collection == "" {
		return nil, fmt.Errorf("Collection name is required")
	}

	if searchMethod == "" {
		return nil, fmt.Errorf("Search method is required")
	}

	nsOpts := &NamespaceOptions{
		namespace: "",
	}

	for _, opt := range opts {
		opt(nsOpts)
	}

	result := hostRecomputeIndex(&collection, &nsOpts.namespace, &searchMethod)

	if result == nil {
		return nil, fmt.Errorf("Failed to recompute")
	}

	return result, nil
}

func ComputeDistance(collection, searchMethod, key1, key2 string, opts ...NamespaceOption) (*CollectionSearchResultObject, error) {
	if collection == "" {
		return nil, fmt.Errorf("Collection name is required")
	}

	if searchMethod == "" {
		return nil, fmt.Errorf("Search method is required")
	}

	if key1 == "" {
		return nil, fmt.Errorf("Key1 is required")
	}

	if key2 == "" {
		return nil, fmt.Errorf("Key2 is required")
	}

	nsOpts := &NamespaceOptions{
		namespace: "",
	}

	for _, opt := range opts {
		opt(nsOpts)
	}

	result := hostComputeDistance(&collection, &nsOpts.namespace, &searchMethod, &key1, &key2)

	if result == nil {
		return nil, fmt.Errorf("Failed to compute distance")
	}

	return result, nil
}

func GetText(collection, key string, opts ...NamespaceOption) (string, error) {
	if collection == "" {
		return "", fmt.Errorf("Collection name is required")
	}

	if key == "" {
		return "", fmt.Errorf("Key is required")
	}

	nsOpts := &NamespaceOptions{
		namespace: "",
	}

	for _, opt := range opts {
		opt(nsOpts)
	}

	result := hostGetText(&collection, &nsOpts.namespace, &key)

	if result == nil {
		return "", fmt.Errorf("Failed to get text for key")
	}

	return *result, nil
}

func GetTexts(collection string, opts ...NamespaceOption) (map[string]string, error) {
	if collection == "" {
		return nil, fmt.Errorf("Collection name is required")
	}

	nsOpts := &NamespaceOptions{
		namespace: "",
	}

	for _, opt := range opts {
		opt(nsOpts)
	}

	result := hostDumpTexts(&collection, &nsOpts.namespace)

	if result == nil {
		return nil, fmt.Errorf("Failed to get texts")
	}

	return *result, nil
}

func GetNamespaces(collection string) ([]string, error) {
	if collection == "" {
		return nil, fmt.Errorf("Collection name is required")
	}

	result := hostGetNamespaces(&collection)

	if result == nil {
		return nil, fmt.Errorf("Failed to get namespaces")
	}

	return *result, nil
}

func GetVector(collection, searchMethod, key string, opts ...NamespaceOption) ([]float32, error) {
	if collection == "" {
		return nil, fmt.Errorf("Collection name is required")
	}

	if searchMethod == "" {
		return nil, fmt.Errorf("Search method is required")
	}

	if key == "" {
		return nil, fmt.Errorf("Key is required")
	}

	nsOpts := &NamespaceOptions{
		namespace: "",
	}

	for _, opt := range opts {
		opt(nsOpts)
	}

	result := hostGetVector(&collection, &nsOpts.namespace, &searchMethod, &key)

	if result == nil {
		return nil, fmt.Errorf("Failed to get vector for key")
	}

	return *result, nil
}

func GetLabels(collection, key string, opts ...NamespaceOption) ([]string, error) {
	if collection == "" {
		return nil, fmt.Errorf("Collection name is required")
	}

	if key == "" {
		return nil, fmt.Errorf("Key is required")
	}

	nsOpts := &NamespaceOptions{
		namespace: "",
	}

	for _, opt := range opts {
		opt(nsOpts)
	}

	result := hostGetLabels(&collection, &nsOpts.namespace, &key)

	if result == nil {
		return []string{}, nil
	}

	return *result, nil
}
